### ISSUE TITLE
Hook CPUID

### ISSUE THREAD

**Issue by Tron0xHex** [unicorn-engine/unicorn#1437]
Hello, i'm trying to hook cpuid instruction, but unicorn gives me an error (UC_ERR_HOOK), my code below.

```
def hook_cpuid(uc: Uc, user_data):
    eax = uc.reg_read(UC_X86_REG_EAX)
    eax, ebx, ecx, edx = cpuid.cpuid(eax)

    uc.reg_write(UC_X86_REG_EAX, eax)
    uc.reg_write(UC_X86_REG_EBX, ebx)
    uc.reg_write(UC_X86_REG_ECX, ecx)
    uc.reg_write(UC_X86_REG_EDX, edx)


mu.hook_add(UC_HOOK_INSN, hook_cpuid, None, 1, 0, UC_X86_INS_CPUID)
```

Can you help me please ?

**Comment by Tron0xHex** [comment]
@wtdcode Sorry for the delay, here is an example:

```
from unicorn import *
from unicorn.x86_const import *

CODE = b'\x40\x0F\xA2'  # INC EAX; CPUID
ADDRESS = 0x1000000


def hook_cpuid(uc: Uc, _):
    uc.reg_write(UC_X86_REG_EAX, 0x007)


def main():
    mu = Uc(UC_ARCH_X86, UC_MODE_32)

    mu.mem_map(ADDRESS, 1024 * 1024)
    mu.mem_write(ADDRESS, CODE)

    mu.hook_add(UC_HOOK_INSN, hook_cpuid, None, 1, 0, UC_X86_INS_CPUID)
    mu.emu_start(ADDRESS, ADDRESS + len(CODE))
    
    print(f"EAX REG: 0x{mu.reg_read(UC_X86_REG_EAX):X}")
    print(f"EBX REG: 0x{mu.reg_read(UC_X86_REG_EBX):X}")
    print(f"ECX REG: 0x{mu.reg_read(UC_X86_REG_ECX):X}")
    print(f"EDX REG: 0x{mu.reg_read(UC_X86_REG_EDX):X}")


if __name__ == "__main__":
    main()
```

**Comment by elicn** [comment]
Browsing the code shows that the `UC_HOOK_INSN` hook type support is limited to only 3 instructions:
- `UC_X86_INS_IN`
- `UC_X86_INS_OUT`
- `UC_X86_INS_SYSCALL`

Specifying any other instruction would result in an error.

See:
- https://github.com/unicorn-engine/unicorn/blob/master/bindings/python/unicorn/unicorn.py#L560
- https://github.com/unicorn-engine/unicorn/blob/master/bindings/java/unicorn_Unicorn.c#L549

**Comment by danyhm** [comment]
I'm also looking for a program that supports CPUID instruction hooking.

I know it's possible with micro-hypervisors but those require signed .sys files for windows. 
So is it possible to add support for CPUID instruction for unicorn?

**Comment by wtdcode** [comment]
@danyhm Yes, we would add support for CPUID after #1217 releases. If not, please ping me then.

**Comment by danyhm** [comment]
@wtdcode sure thanks! i hope that happens soon!

**Comment by aquynh** [comment]
we released Unicorn2 beta to the public.
check it out the "dev" branch.

**Comment by wtdcode** [comment]
Fixed in 9d8a309fbf688eb20eac224d465b980afbb052d9
